read/readtags.c

```diff
@@ -355,20 +355,38 @@ static void parseExtensionFields (tagFile *const file, tagEntry *const entry,
 			else
 			{
 				const char *key = field;
-				const char *value = colon + 1;
-				*colon = '\0';
-				if (strcmp (key, "kind") == 0)
-					entry->kind = value;
-				else if (strcmp (key, "file") == 0)
-					entry->fileScope = 1;
-				else if (strcmp (key, "line") == 0)
-					entry->address.lineNumber = atol (value);
+				if (colon - key == 4 && key[1] == 'i')
+				{
+					if (key[3] == 'e')
+					{
+						if (key[0] == 'f' &&
+							key[2] == 'l')
+							/* file: */
+							entry->fileScope = 1;
+						else if (key[0] == 'l' &&
+								 key[2] == 'n')
+							/* line: */
+							entry->address.lineNumber = atol (colon + 1);
+						else
+							goto minorField;
+
+					}
+					else if (key[0] == 'k' &&
+							 key[2] == 'n' &&
+							 key[3] == 'd')
+						/* kind: */
+						entry->kind = colon + 1;
+					else
+						goto minorField;
+				}
```

Comment:
I think you need to ask yourself whether the extra performance is worth this kind of code, that'll be a nightmare to maintain (IMO; I can't imagine the pain of adding a new case).

I'd rather go halfway myself, keep the initial computation for the length, but instead of convoluted checking for the key simply use `if (key[0] == 'f' && key[1] == 'i' && key[2] == 'l' && key[3] == 'e')` which will likely be faster than strcmp, maybe even very close to what you have right now, but less complicated.  If that works, you could even make it more readable with a small macro e.g. like `#define CMP4(a, b) ((a)[0] == (b)[0] && (a)[1] == (b)[1] && (a)[2] == (b)[2] && (a)[3] == (b)[3])` and use `if (CMP4(key, "file"))`.

I'd also keep the `value` variable that is used in all paths, unless it's really something the compiler chokes on -- but that's be crazy of it. Make it a constant if you'd like though.

Anyway, all this is not checked for speed, but IMO not all speed gains are worth it if the code becomes too problematic.

Reply:
This said, I'd be kind of surprised that a good compiler with optimizations enabled wouldn't do something similar to my `CMP4()` all by itself, and could do that better using specific instruction sets…
Have you tried simply guarding with the length check and keeping the `strcmp()`s?

Reply:
> I'd rather go halfway myself, keep the initial computation for the length

I see. My understanding is that you are talking about:
```
if (colon - key == 4
```
.

Reply:
Yes.

But remember that I'm just looking at the code, I didn't profile anything, so don't take my word as gospel :wink:

Reply:
All you wrote are correct as new result of profiling shows.